cpp/src/prims/detail/nbr_intersection.cuh

```diff
@@ -436,33 +524,63 @@ struct strided_accumulate_t {
   }
 };
 
-template <typename vertex_t>
+template <typename vertex_t, typename EdgeProperty_t>
 struct gatherv_indices_t {
   size_t output_size{};
   int minor_comm_size{};
 
   raft::device_span<size_t const> gathered_intersection_offsets{};
   raft::device_span<vertex_t const> gathered_intersection_indices{};
   raft::device_span<size_t const> combined_nbr_intersection_offsets{};
-
   raft::device_span<vertex_t> combined_nbr_intersection_indices{};
 
+  std::optional<raft::device_span<EdgeProperty_t const>> gathered_nbr_intersection_properties0{
+    std::nullopt};
+  std::optional<raft::device_span<EdgeProperty_t const>> gathered_nbr_intersection_properties1{
+    std::nullopt};
+
+  std::optional<raft::device_span<EdgeProperty_t>> combined_nbr_intersection_properties0{
+    std::nullopt};
+  std::optional<raft::device_span<EdgeProperty_t>> combined_nbr_intersection_properties1{
+    std::nullopt};
+
   __device__ void operator()(size_t i) const
   {
     auto output_offset = combined_nbr_intersection_offsets[i];
 
     // FIXME: this can lead to thread-divergence with a mix of high-degree and low-degree vertices
     // in a single warp (better optimize if this becomes a performance bottleneck)
 
+    auto zipped_gathered_begin =
+      thrust::make_zip_iterator(thrust::make_tuple(gathered_intersection_indices.begin(),
+                                                   gathered_nbr_intersection_properties0->begin(),
+                                                   gathered_nbr_intersection_properties1->begin()));
+
+    auto zipped_combined_begin =
+      thrust::make_zip_iterator(thrust::make_tuple(combined_nbr_intersection_indices.begin(),
+                                                   combined_nbr_intersection_properties0->begin(),
+                                                   combined_nbr_intersection_properties1->begin()));
```

Comment:
`gathered_nbr_intersection_properties0|1` and `combined_nbr_intersection_indices0|1` is not always valid. And IIRC, -> on std::optional throws an exception when the object is invalid, but the device code cannot handle exceptions. Better not use -> for optional on device.